### PR TITLE
SVM classifier abstraction

### PR DESCRIPTION
**Pull request by Regis128**

### CODE REVIEW

File: src/Spectre.libClassifier/IClassifier.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /// </summary>
+    class IClassifier
+    {
+    public:
+        using LabeledDataset = const libDataset::IReadOnlyDataset<Observation, Label, Empty>&;
+        /// <summary>
+        /// Try to fit dataset in fitness function.
+        /// </summary>
+        /// <param name="dataset">The dataset.</param>
+        /// <returns>void</returns>
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/IClassifier.h]: This tag should be removed.

File: src/Spectre.libClassifier/IClassifier.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /// <summary>
+    /// Classifier base class.
+    /// </summary>
+    class IClassifier
+    {
+    public:
+        using LabeledDataset = const libDataset::IReadOnlyDataset<Observation, Label, Empty>&;
+        /// <summary>
+        /// Try to fit dataset in fitness function.
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/IClassifier.h]: It fits the classifier to the data. There's no explicit *fitness function* term on the outside.

File: src/Spectre.libClassifier.Tests/SvmTest.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <gtest/gtest.h>
+#include "Spectre.libClassifier/Types.h"
+#include "Spectre.libClassifier/OpenCvDataset.h"
+#include "Spectre.libClassifier/Svm.h"
+
+namespace
+{
+    using namespace Spectre::libClassifier;
+
```
Comment (gmrukwa) [inline src/Spectre.libClassifier.Tests/SvmTest.cpp]: Number of support vectors has not been tested yet.

File: src/Spectre.libClassifier/Svm.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        m_Svm->predict(data.getMatData(), predictionOutput);
+        std::vector<Label> labels(numberOfObservations, 0);
+        cv::Mat labelsOutput(numberOfObservations, 1, CV_LABEL_TYPE, labels.data());
+        predictionOutput.convertTo(labelsOutput, CV_LABEL_TYPE);
+        return labels;
+    }
+
+    unsigned int Svm::GetNumberOfSupportVectors() const
+    {
+        return m_Svm->getSupportVectors().rows;
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/Svm.cpp]: This is not necessarily valid for SVC SVM.

File: src/Spectre.libClassifier/Svm.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    {
+    public:
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Svm"/> class.
+        /// </summary>
+        /// <param name="iterationsLimit">The iterations limit.</param>
+        /// <param name="tolerance">The tolerance.</param>
+        explicit Svm(unsigned int iterationsLimit = 100, double tolerance = 1e-6);
+        /// <summary>
+        /// Try to fit dataset in fitness function.
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/Svm.h]: Fix, similarly as in the IClassifier.

File: src/Spectre.libClassifier/Svm.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        m_Svm->train(data.getMatData(), cv::ml::ROW_SAMPLE, data.getMatLabels());
+    }
+
+    std::vector<Label> Svm::Predict(LabeledDataset dataset) const
+    {
+        const auto& data = asSupported(dataset);
+        const auto numberOfObservations = static_cast<unsigned>(data.getMatData().rows);
+        std::vector<float> predictions(numberOfObservations, 0);
+        cv::Mat predictionOutput(numberOfObservations, 1, CV_TYPE, predictions.data());
+        m_Svm->predict(data.getMatData(), predictionOutput);
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/Svm.cpp]: What if user tries to predict without fitting previously?

File: src/Spectre.libClassifier/IClassifier.h (hunk 2 above)
Comment (Regis128) [reply src/Spectre.libClassifier/IClassifier.h]: done

File: src/Spectre.libClassifier/IClassifier.h (hunk 1 above)
Comment (Regis128) [reply src/Spectre.libClassifier/IClassifier.h]: done

File: src/Spectre.libClassifier/Svm.h (hunk 5 above)
Comment (Regis128) [reply src/Spectre.libClassifier/Svm.h]: done

File: src/Spectre.libClassifier.Tests/SvmTest.cpp (hunk 3 above)
Comment (Regis128) [reply src/Spectre.libClassifier.Tests/SvmTest.cpp]: done

File: src/Spectre.libClassifier/Svm.cpp (hunk 4 above)
Comment (Regis128) [reply src/Spectre.libClassifier/Svm.cpp]: For SVM I cannot see other way to get it.

File: src/Spectre.libClassifier/Svm.cpp (hunk 6 above)
Comment (Regis128) [reply src/Spectre.libClassifier/Svm.cpp]: done

File: src/Spectre.libClassifier.Tests/SvmTest.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    TEST_F(SvmTest, get_support_vector_number_test)
+    {
+        Svm svm;
+        svm.Fit(m_Data);
+        EXPECT_EQ(svm.GetNumberOfSupportVectors(), 1);
+    }
+
+    TEST_F(SvmTest, predicts_without_train)
```
Comment (gmrukwa) [inline src/Spectre.libClassifier.Tests/SvmTest.cpp]: Rename it to be more accurate, like `throws_...`.

**Pull request merged**